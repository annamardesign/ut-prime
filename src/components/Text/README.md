# Button

## How to use

```jsx
import Text from 'ut-front-devextreme/core/Text';

<Text onClick={handlerOnClick}>Hello Text</Text>;
```

## Props

- **className** - class applied to root `Text` html element.

| propName  | propType | defaultValue | isRequired |
| --------- | -------- | ------------ | ---------- |
| className | string   |              | no         |
