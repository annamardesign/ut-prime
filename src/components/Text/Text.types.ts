import React from 'react';

export interface Props extends React.HTMLAttributes<HTMLSpanElement> {
    prefix?: string;
    params?: {};
}

export type ComponentProps = React.FC<Props>
