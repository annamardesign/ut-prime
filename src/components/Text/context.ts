import React from 'react';

interface contextType {
    language?: string;
    translate?: (id: string, text:string, language: string) => string
}
const defaultContext: contextType = {};
export default React.createContext(defaultContext);
