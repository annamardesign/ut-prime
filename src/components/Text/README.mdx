# Text
import { Canvas, Story, ArgsTable } from "@storybook/addon-docs";

Component to show text, that can be optionally translated to other languages.

## How to use

```jsx
import Text from 'ut-prime/core/Text';

<Text>Text content, which can be translated</Text>
```

The component will look for а language and а translation
function in the React Context and use them
to translate the embedded text.

<Canvas>
    <Story id="text--basic" />
</Canvas>

## Properties

<ArgsTable story="."/>
