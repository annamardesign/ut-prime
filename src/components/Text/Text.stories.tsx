import React from 'react';
import { withReadme } from 'storybook-readme';

// @ts-ignore: md file and not a module
import README from './README.md';
import Text from './index';

export default {
    title: 'Text',
    component: Text,
    decorators: [withReadme(README)],
    args: {
        state: {
        }
    }
};

export const Basic: React.FC<{}> = () =>
    <div className='p-component'><Text>Text content, which can be translated</Text></div>;
