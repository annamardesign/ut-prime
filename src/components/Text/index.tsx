import React, { useContext } from 'react';

import Context from '../Context';

import { StyledType } from './Text.types';

const TOKEN = /\${([\w]*)}/g;
/**
 * A very simple template scheme:
 * template = 'This ${item} costs ${price}'
 * params = {item: 'boots', price: '100'})
 * result: 'This boots costs 100, John'
 */
function applyTemplate(template, params) {
    if (typeof template === 'string' && params) {
        return template.replace(TOKEN, function(wholeMatch, key) {
            return (key in params &&
                    typeof params[key] !== 'object' &&
                    typeof params[key] !== 'function' &&
                    typeof params[key] !== 'undefined')
                ? params[key]
                : wholeMatch;
        });
    } else {
        // We cannot do any processing - just return the original
        return template;
    }
}

const Text: StyledType = ({ params, prefix, children }) => {
    if (typeof children !== 'string') return <>children</>;
    let template = children;
    const {translate, language} = useContext(Context);
    if (typeof translate === 'function') {
        const text = (prefix ? [prefix, children] : [children]).join('>');
        // Translate the template
        template = translate(text, language);
        if (template === text) {
            template = children;
        }
    }
    return <span>{applyTemplate(template, params)}</span>;
};

export default Text;
