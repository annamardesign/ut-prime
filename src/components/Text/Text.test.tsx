import React from 'react';

import { render } from '../test';
import Text from './index';

describe('<Text />', () => {
    it('render equals snapshot', () => {
        const { getByTestId } = render(<Text>text</Text>);
        expect(getByTestId('ut-front-test')).toMatchSnapshot();
    });
});
