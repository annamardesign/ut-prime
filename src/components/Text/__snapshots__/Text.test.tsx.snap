// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  <span>
    text
  </span>
</div>
`;
